typescript/sdk/src/token/RadixWarpRouteReader.ts

```diff
@@ -0,0 +1,162 @@
+   *
+   * @param warpRouteAddress - The address of the Hyperlane ERC20 router contract.
+   * @returns The configuration for the Hyperlane ERC20 router.
+   *
+   */
+  async deriveWarpRouteConfig(
+    warpRouteAddress: Address,
+  ): Promise<DerivedTokenRouterConfig> {
+    // Derive the config type
+    const type = await this.deriveTokenType(warpRouteAddress);
+    const baseMetadata = await this.fetchMailboxClientConfig(warpRouteAddress);
+    const tokenConfig = await this.fetchTokenConfig(type, warpRouteAddress);
+    const remoteRouters = await this.fetchRemoteRouters(warpRouteAddress);
+    const destinationGas = await this.fetchDestinationGas(warpRouteAddress);
+
+    return {
+      ...baseMetadata,
+      ...tokenConfig,
+      remoteRouters,
+      destinationGas,
+      type,
+    } as DerivedTokenRouterConfig;
+  }
+
+  /**
+   * Derives the token type for a given Warp Route address using specific methods
+   *
+   * @param warpRouteAddress - The Warp Route address to derive the token type for.
+   * @returns The derived token type, which can be one of: collateralVault, collateral, native, or synthetic.
+   */
+  async deriveTokenType(warpRouteAddress: Address): Promise<TokenType> {
+    const token = await this.findToken(warpRouteAddress);
+
+    switch (token.token_type) {
+      case RadixTokenTypes.COLLATERAL:
+        return TokenType.collateral;
+      case RadixTokenTypes.SYNTHETIC:
+        return TokenType.synthetic;
+      default:
+        throw new Error(
+          `Radix unknown token type on token contract ${warpRouteAddress}: ${token.token_type}`,
+        );
+    }
+  }
+
+  /**
+   * Fetches the base metadata for a Warp Route contract.
+   *
+   * @param routerAddress - The address of the Warp Route contract.
+   * @returns The base metadata for the Warp Route contract, including the mailbox, owner, hook, and ism.
+   */
+  async fetchMailboxClientConfig(
+    routerAddress: Address,
+  ): Promise<MailboxClientConfig> {
+    const token = await this.findToken(routerAddress);
+
+    const config: MailboxClientConfig = {
+      mailbox: token.mailbox,
+      owner: token.owner,
+    };
+
+    if (token.ism) {
+      const derivedIsm = await this.ismReader.deriveIsmConfig(token.ism);
+      config.interchainSecurityModule = derivedIsm;
+    }
+
+    return config;
+  }
+
+  /**
+   * Fetches the metadata for a token address.
+   *
+   * @param warpRouteAddress - The address of the token.
+   * @returns A partial ERC20 metadata object containing the token name, symbol, total supply, and decimals.
+   * Throws if unsupported token type
+   */
+  async fetchTokenConfig(
+    type: TokenType,
+    warpRouteAddress: Address,
+  ): Promise<HypTokenConfig> {
+    const token = await this.findToken(warpRouteAddress);
+
+    return {
+      type,
+      token: warpRouteAddress,
+      name: token.name,
+      symbol: token.symbol,
+      decimals: token.divisibility,
+    } as HypTokenConfig;
```

Comment:
once again can we try to reduce type assertions as much as possible? if there are fields from the token metadata that might be undefined but are required by the `HypTokenConfig` type we should assert that they exist instead of silencing the error with a type assertion.

Reply:
there is lots of evm specific stuff expected here which will never be used in cosmos or radix